File: src/data/o1/Merge_k_Sorted_Lists/Merge_k_Sorted_Lists.java
import java.util.PriorityQueue;
import java.util.Comparator;

class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists == null || lists.length == 0) return null;
        
        PriorityQueue<ListNode> heap = new PriorityQueue<>(new Comparator<ListNode>() {
            public int compare(ListNode a, ListNode b) {
                return a.val - b.val;
            }
        });
        
        for (ListNode node : lists) {
            if (node != null) {
                heap.offer(node);
            }
        }
        
        ListNode dummy = new ListNode(0);
        ListNode current = dummy;
        
        while (!heap.isEmpty()) {
            ListNode minNode = heap.poll();
            current.next = minNode;
            current = current.next;
            if (minNode.next != null) {
                heap.offer(minNode.next);
            }
        }
        
        return dummy.next;
    }
}

public class Merge_k_Sorted_Lists {
// Definition for singly-linked list.
class ListNode {
    int val;
    ListNode next;

    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

// Utility function to convert a list to a linked list
public static ListNode listToLinkedList(int[] lst) {
    ListNode dummy = new ListNode();
    ListNode current = dummy;
    for (int val : lst) {
        current.next = new ListNode(val);
        current = current.next;
    }
    return dummy.next;
}

// Utility function to convert a linked list to a list
public static int[] linkedListToList(ListNode node) {
    List<Integer> result = new ArrayList<>();
    while (node != null) {
        result.add(node.val);
        node = node.next;
    }
    return result.stream().mapToInt(i -> i).toArray();
}

// Method to run all test cases
public static void runTests(Solution solution) {
    // Test case 1: Merging three sorted lists
    ListNode[] lists1 = {
        listToLinkedList(new int[]{1, 4, 5}),
        listToLinkedList(new int[]{1, 3, 4}),
        listToLinkedList(new int[]{2, 6})
    };
    int[] expected_result1 = {1, 1, 2, 3, 4, 4, 5, 6};
    ListNode result1 = solution.mergeKLists(lists1);
    assert Arrays.equals(linkedListToList(result1), expected_result1) : "Test case 1 failed. Expected " + Arrays.toString(expected_result1);

    // Test case 2: Single empty list
    ListNode[] lists2 = {null};
    int[] expected_result2 = {};
    ListNode result2 = solution.mergeKLists(lists2);
    assert Arrays.equals(linkedListToList(result2), expected_result2) : "Test case 2 failed. Expected " + Arrays.toString(expected_result2);

    // Test case 3: All lists are empty
    ListNode[] lists3 = {null, null, null};
    int[] expected_result3 = {};
    ListNode result3 = solution.mergeKLists(lists3);
    assert Arrays.equals(linkedListToList(result3), expected_result3) : "Test case 3 failed. Expected " + Arrays.toString(expected_result3);

    // Test case 4: Mixed empty and non-empty lists
    ListNode[] lists4 = {
        listToLinkedList(new int[]{}),
        listToLinkedList(new int[]{1, 3, 4}),
        null,
        listToLinkedList(new int[]{2, 6, 7})
    };
    int[] expected_result4 = {1, 2, 3, 4, 6, 7};
    ListNode result4 = solution.mergeKLists(lists4);
    assert Arrays.equals(linkedListToList(result4), expected_result4) : "Test case 4 failed. Expected " + Arrays.toString(expected_result4);

    // Test case 5: All lists with single element
    ListNode[] lists5 = {
        listToLinkedList(new int[]{1}),
        listToLinkedList(new int[]{0}),
        listToLinkedList(new int[]{5})
    };
    int[] expected_result5 = {0, 1, 5};
    ListNode result5 = solution.mergeKLists(lists5);
    assert Arrays.equals(linkedListToList(result5), expected_result5) : "Test case 5 failed. Expected " + Arrays.toString(expected_result5);

    // Test case 6: No lists provided
    ListNode[] lists6 = {};
    int[] expected_result6 = {};
    ListNode result6 = solution.mergeKLists(lists6);
    assert Arrays.equals(linkedListToList(result6), expected_result6) : "Test case 6 failed. Expected " + Arrays.toString(expected_result6);
}

    public static void main(String[] args) {
        Solution solution = new Solution();
        runTests(solution);   // Ensure your test cases are implemented inside runTests
    }
}

    

File: src/data/4o/Merge_k_Sorted_Lists/Merge_k_Sorted_Lists.cpp
#include <queue>
#include <vector>
using namespace std;


struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        // Lambda comparator for the priority queue (min-heap)
        auto compare = [](ListNode* a, ListNode* b) {
            return a->val > b->val;
        };

        priority_queue<ListNode*, vector<ListNode*>, decltype(compare)> pq(compare);

        // Push the head of each list into the priority queue
        for (auto list : lists) {
            if (list) {
                pq.push(list);
            }
        }

        ListNode* dummy = new ListNode(0);
        ListNode* tail = dummy;

        // Merge the lists by always taking the smallest element from the heap
        while (!pq.empty()) {
            ListNode* node = pq.top();
            pq.pop();
            tail->next = node;
            tail = tail->next;

            if (node->next) {
                pq.push(node->next);
            }
        }

        return dummy->next;
    }
};

//  Test cases for the C++ solution
#include <cassert>

// Helper function to convert vector to linked list
ListNode* list_to_linkedlist(const std::vector<int>& lst) {
    ListNode* dummy = new ListNode();
    ListNode* current = dummy;
    for (int val : lst) {
        current->next = new ListNode(val);
        current = current->next;
    }
    return dummy->next;
}

// Helper function to convert linked list to vector
std::vector<int> linkedlist_to_list(ListNode* node) {
    std::vector<int> result;
    while (node) {
        result.push_back(node->val);
        node = node->next;
    }
    return result;
}

// Test cases for the C++ solution
void testSolution(Solution& solution) {
    // Test case 1: Merging three sorted lists
    std::vector<ListNode*> lists = {
        list_to_linkedlist({1, 4, 5}),
        list_to_linkedlist({1, 3, 4}),
        list_to_linkedlist({2, 6})
    };
    std::vector<int> expected_result = {1, 1, 2, 3, 4, 4, 5, 6};
    ListNode* result = solution.mergeKLists(lists);
    assert(linkedlist_to_list(result) == expected_result && "Test case 1 failed.");

    // Test case 2: Single empty list
    lists = {nullptr};
    expected_result = {};
    result = solution.mergeKLists(lists);
    assert(linkedlist_to_list(result) == expected_result && "Test case 2 failed.");

    // Test case 3: All lists are empty
    lists = {nullptr, nullptr, nullptr};
    expected_result = {};
    result = solution.mergeKLists(lists);
    assert(linkedlist_to_list(result) == expected_result && "Test case 3 failed.");

    // Test case 4: Mixed empty and non-empty lists
    lists = {
        list_to_linkedlist({}),
        list_to_linkedlist({1, 3, 4}),
        nullptr,
        list_to_linkedlist({2, 6, 7})
    };
    expected_result = {1, 2, 3, 4, 6, 7};
    result = solution.mergeKLists(lists);
    assert(linkedlist_to_list(result) == expected_result && "Test case 4 failed.");

    // Test case 5: All lists with single element
    lists = {
        list_to_linkedlist({1}),
        list_to_linkedlist({0}),
        list_to_linkedlist({5})
    };
    expected_result = {0, 1, 5};
    result = solution.mergeKLists(lists);
    assert(linkedlist_to_list(result) == expected_result && "Test case 5 failed.");

    // Test case 6: No lists provided
    lists = {};
    expected_result = {};
    result = solution.mergeKLists(lists);
    assert(linkedlist_to_list(result) == expected_result && "Test case 6 failed.");
}

int main() {
    Solution solution;
    for (int i = 0; i < 655360; ++i) {
        testSolution(solution);  
    }
    return 0;
}
